src/internet/model/tcp-socket-base.cc

```diff
@@ -4156,4 +4098,592 @@ RttHistory::RttHistory (const RttHistory& h)
+      NS_LOG_ERROR ("Estimation type Padhye not yet implemented!");
+      break;
+    case FairShareEstimationType::Average:
+      NS_LOG_ERROR ("EstimateFairShareCalc() can only be used with estimation types Mathis and Padhye!");
+      break;
+    case FairShareEstimationType::Model:
+      NS_LOG_ERROR ("EstimateFairShareCalc() can only be used with estimation types Mathis and Padhye!");
+      break;
+    default:
+      NS_LOG_ERROR ("Invalid fair share estimation type!");
+    }
+  return fairTput;
+}
+
+double
+TcpSocketBase::EstimateFairShareAverage (Ptr<Packet> p)
+{
+  NS_LOG_FUNCTION (this << p);
+  double totalTput = 0;
+  for (const auto& socket : sockets)
+    {
+      totalTput += socket->GetStats ().tputMbps;
+    }
+  return totalTput / sockets.size ();
+  return 0;
+}
+
+Time
+TcpSocketBase::EstimateAckPeriodCalc (double actualTput, double targetTput,
+                                      Time rtt)
+{
+  NS_LOG_FUNCTION (this << actualTput << targetTput << rtt);
+  return rtt * sqrt (actualTput / targetTput) * 0.25;
+}
+
+Time
+TcpSocketBase::EstimateAckPeriodModel (double targetTputMbps)
+{
+  NS_LOG_FUNCTION (this << targetTputMbps);
+  // std::vector<torch::jit::IValue> input ({scale(targetTputMbps * 1e6)});
+  // The net takes (current tput (Gbps), target tput (Gbps), RTT (ss))
+  // dummy.push_back(torch::tensor({stats.tputMbps/1e3, targetTputMbps/1e3, stats.avgLat.GetSeconds()*2.}));
+  //input.push_back(torch::tensor({target}));
+  // periodTensor.print();
+  // Returns ack period (s)
+  //auto output = periodTensor.item().to<double>();
+
+  // TODO: This scaling assumes a single input and output feature.
+  double out = m_net->forward (std::vector<torch::jit::IValue> ({
+        Scale (targetTputMbps * 1e6,
+               std::get<0> (std::get<0> (m_scaleParams)[0]),
+               std::get<1> (std::get<0> (m_scaleParams)[0]), 0, 1)}))
+    .toTensor ().item ().to<double> ();
+  double out_unscaled = Scale (out, 0, 1,
+                               std::get<0> (std::get<1> (m_scaleParams)[0]),
+                               std::get<1> (std::get<1> (m_scaleParams)[0]));
+  NS_LOG_INFO ("Raw output: " << out << ", unscaled: " << out_unscaled);
+  return MicroSeconds (1. / out_unscaled);
+}
+
+void
+TcpSocketBase::ScheduleAckPacket (Ptr<TcpL4Protocol> tcp, Ptr<Packet> p,
+                                  TcpHeader header, Ipv4Address localaddr,
+                                  Ipv4Address peeraddr,
+                                  Ptr<NetDevice> boundnetdevice)
+{
+  NS_LOG_FUNCTION (this << tcp << p << header << localaddr << peeraddr <<
+                   boundnetdevice);
+  // If the ACK period is 0 or the next time at which an ACK should be sent is
+  // in the past, then send the ACK immediately. Otherwise, enqueue it for
+  // tranmission later.
+  if (m_ackPeriod == 0 || m_prevAckTime + m_ackPeriod < Simulator::Now ())
+    {
+      // TODO: Debug why this assertion is sometimes invalid.
```

Comment:
I think it's because the measured throughput goes from above 1.2x to lower 1.2x. For the previous packet, ACK pacing is still on and pendingAcks queue is not empty. Once the measured throughput is lower than 1.2x, m_ackPeriod == 0, but there're still packets inside the queue.